core/src/main/java/org/jruby/javasupport/ext/JavaExtensions.java

```diff
@@ -0,0 +1,43 @@
+package org.jruby.javasupport.ext;
+
+import org.jruby.Ruby;
+import org.jruby.RubyModule;
+
+import java.util.Map;
+import java.util.WeakHashMap;
+import java.util.function.Consumer;
+
+/**
+ * Lazy Java class extensions initialization.
+ *
+ * @note Internal API
+ * @author kares
+ */
+public class JavaExtensions {
+
+    private static final boolean IMMEDIATE = false;
+
+    private static final JavaExtensions INSTANCE = new JavaExtensions();
```

Comment:
This should not be static; a better place would be for it to live on the javasupport Java instance attached to the runtime.

It is weak on the value, but it also anchors a class as its key. The classes we currently extend are system-level classes, but this may be used in the future for classes loaded as part of JRuby or within extensions (e.g. jruby-openssl) that load into their own classloaders, and this reference will prevent them from being collected.

It also is not compatible with multiple JRuby instances in the same JVM, since only the first one will register (and eventually apply) an extension. The others will race to insert the extension or remove it when applying that extension.

Reply:
Oh somehow my review used an older version that was still static. I see you have fixed that, thanks!